design-proposals/klip-59-json-functions.md

```diff
@@ -0,0 +1,244 @@
+representing the top-level keys and values. Returns `NULL` if the string can't be interpreted as a
+JSON object, i.e. it is `NULL` or it does not contain valid JSON, or the JSON value is not an object.
+
+#### Examples
+
+```
+json_records("{\"a\": \"abc\", \"b\": { \"c\": \"a\" }, \"d\": 1}") // returns [["a", "\"abc\""], ["b", "{ \"c\": \"a\" }"], ["d", "1"]]
+json_records("{}") // returns []
+json_records("[]") // returns NULL
+json_records("") // returns NULL
+json_records("123") // returns NULL
+json_records("abc") // returns NULL
+json_records(NULL) // returns NULL
+```
+
+#### Rejected alternatives
+
+Remove redundant quotes around values. For example:
+
+```
+json_records("{\"a\": \"abc\"}) // returns [["a", "abc"]]
+```
+
+This makes the output less awkward, but removes the type information in certain cases. In the following
+case it is not possible to derive whether `1` is a string or a number:
+
+```
+json_records("{\"a\": \"1\"}) // returns [["a", "1"]]
+```
+
+### json_concat
+
+```
+json_concat(json_string, json_string)
+```
+
+Given 2 strings, parse them as JSON values and return a string representing their concatenation.
+Concatenation rules are identical to PostgreSQL's [|| operator](https://www.postgresql.org/docs/14/functions-json.html):
+* If both strings deserialize into JSON objects, then return an object with a union of the input key,
+  taking values from the second object in the case of duplicates.
+* If at least one of the deserialized values is not an object, then convert non-array inputs to a
+  single-element array and return the result of array concatenation.
+* If at least one of the input strings is `NULL` or can't be deserialized as JSON, then return `NULL`.
+
+Akin to PostgreSQL's `||` operator, this function merges only top-level object keys or arrays.
+
+#### Examples
+
+```
+json_concat("{\"a\": 1}", "{\"b\": 2}") // returns "{\"a\": 1, \"b\": 2}"
+json_concat("{\"a\": {\"3\": 4}}", "{\"a\": {\"5\": 6}}") // returns "{\"a\": {\"5\": 6}}"
+json_concat("{\"a\": {\"5\": 6}}", "{\"a\": {\"3\": 4}}") // returns "{\"a\": {\"3\": 4}}"
+json_concat("{}", "{}") // returns "{}"
+json_concat("[1, 2]", "[3, 4]") // returns "[1,2,3,4]"
+json_concat("[1, [2]]", "[[[3]], [[[4]]]]") // returns "[ 1, [2], [[3]], [[[4]]] ]"
+json_concat("null", "null") // returns [null, null]
+json_concat("[1, 2]", "{\"a\": 1}") // returns "[1, 2, {\"a\": 1}]"
+json_concat("[1, 2]", "3") // returns "[1, 2, 3]"
+json_concat("[3]", "[1, 2]") // returns "[3, 1, 2]"
+json_concat("1", "2") // returns "[1, 2]"
+json_concat("[]", "[]") // returns []
+json_concat("abc", "[1]") // returns NULL
+json_concat(NULL, "[1]") // returns NULL
+```
+
+#### Rejected alternatives
+
+If one argument represents a valid JSON and the other one does not, then return the valid string
+as a result of concatenation.
+
+Given that concatenation of invalid JSON strings is an exceptional situation, I believe that we
```

Comment:
nit: remove `I believe that`

Reply:
Removed in https://github.com/confluentinc/ksql/pull/8550/commits/9672f21f932c16bf173529889608035ce467270c.